File: aoc.main/src/aoc/day01.clj
(ns aoc.day01
  (:require [aoc.util :as util]
            [clojure.string :as str]))

(defn parse-data [input]
  (map #(Integer. %) (str/split-lines input)))

(defn solve-part1 []
  (let [input (-> "1" util/get-input parse-data)]
    (->> input
         (partition 2 1)
         (filter (fn [[x y]] (< x y)))
         count)))

(defn solve-part2 []
  (let [input (-> "1" util/get-input parse-data)]
    (->> input
         (partition 4 1)
         (filter (fn [[a b c d]] (< (+ a b c) (+ b c d))))
         count)))

File: src/day01.rs
use aoc_runner_derive::{aoc, aoc_generator};

#[aoc_generator(day1)]
pub fn parse_input(input: &str) -> Vec<i32> {
    input.lines().map(|s| s.parse::<i32>().unwrap()).collect()
}

#[aoc(day1, part1)]
pub fn short_part1(input: &Vec<i32>) -> usize {
    input.windows(2).filter(|x| x[0] < x[1]).count()
}

#[aoc(day1, part2)]
pub fn short_part2(input: &Vec<i32>) -> usize {
    input
        .windows(4)
        .filter(|x| x[0..3].iter().sum::<i32>() < x[1..4].iter().sum())
        .count()
}
